### DISCUSSION TITLE
how can i get the mean of kprobe func args

### DISCUSSION THREAD

**Discussion by robyn-he** [bpftrace/bpftrace#1790] [General]
![image](https://user-images.githubusercontent.com/44891950/114559451-5339e100-9c9e-11eb-8c3b-fc9857107545.png)
bpftrace -lv 'kprobe:__tcp_select_window' is invalid
how can i get the mean of kprobe func args

**Comment by fbs** [comment]
You can't. Bpftrace doesn't know the type of arguments, it hopes the user does the right thing. So you have to search through the kernel sources. https://elixir.bootlin.com/linux/latest/source is helpful here but make sure you select the right version!

On a modern system we have kfunc support, these use BTF to derive the types and are able to list the arguments. `bpftrace --info` will tell you whether kfuncs are supported or not.

**Reply by robyn-he** [reply]
the kernel sources "__tcp_select_window" has only one arg,but "bpftrace -e 'kprobe:__tcp_select_window { printf ("%d\n",arg5) }'" has 5 args
![image](https://user-images.githubusercontent.com/44891950/114873367-ccb40980-9e2d-11eb-8312-dd136ba0eff7.png)

**Reply by robyn-he** [reply]
how can i get all args mean of kprobe:__tcp_select_window?thanks

**Reply by fbs** [reply]
it doesn't have 5 args.

 Bpftrace _doesn't know_ the amount of args, so it trusts you to do the right thing. Internally argX just grabs the value from the corresponding register (as per calling convention for your CPU architecture, see https://github.com/iovisor/bpftrace/blob/e2fb442afa2cad12f6138976c3a26370e48b9123/src/arch/x86_64.cpp). Registers aren't zerod on function call so any value you see is just a left over from a previous call or use as scratch register.

So in this case only ARG0 is valid, the rest is garbage.
